src/stream/xml_reader.rs

```diff
@@ -1,221 +1,262 @@
+                    }
+                }
+                XmlEvent::Empty(name) => {
+                    match name.local_name() {
+                        b"true" => return Ok(Some(Event::Boolean(true))),
+                        b"false" => return Ok(Some(Event::Boolean(false))),
+
+                        b"array" | b"dict" => {
+                            let owned_name = ElmName::from(name.local_name());
+                            // Open and immediately close a collection element
+                            self.element_stack.push(owned_name.clone());
+                            self.closed_element = Some(owned_name);
+
+                            match name.local_name() {
+                                b"array" => return Ok(Some(Event::StartArray(None))),
+                                b"dict" => return Ok(Some(Event::StartDictionary(None))),
+                                _ => unreachable!(),
+                            }
+                        }
+
                         _ => return Err(self.with_pos(ErrorKind::UnknownXmlElement)),
                     }
                 }
-                Ok(XmlEvent::EndElement { name, .. }) => {
+                XmlEvent::End(name) => {
                     // Check the corrent element is being closed
                     match self.element_stack.pop() {
-                        Some(ref open_name) if &name.local_name == open_name => (),
-                        Some(ref _open_name) => {
+                        Some(open_name) if name.local_name() != open_name.as_ref() => {
                             return Err(self.with_pos(ErrorKind::UnclosedXmlElement))
                         }
                         None => return Err(self.with_pos(ErrorKind::UnpairedXmlClosingTag)),
+                        _ => {}
                     }
 
-                    match &name.local_name[..] {
-                        "array" | "dict" => return Ok(Some(Event::EndCollection)),
-                        "plist" | _ => (),
-                    }
-                }
-                Ok(XmlEvent::EndDocument) => {
-                    if self.element_stack.is_empty() {
-                        return Ok(None);
-                    } else {
-                        return Err(self.with_pos(ErrorKind::UnclosedXmlElement));
+                    match name.local_name() {
+                        b"array" | b"dict" => return Ok(Some(Event::EndCollection)),
+                        b"plist" | _ => (),
                     }
                 }
-
-                Ok(XmlEvent::Characters(c)) => {
-                    if !is_whitespace_str(&c) {
-                        return Err(
-                            self.with_pos(ErrorKind::UnexpectedXmlCharactersExpectedElement)
-                        );
-                    }
+                XmlEvent::Eof if self.element_stack.is_empty() => return Ok(None),
+                XmlEvent::Eof => return Err(self.with_pos(ErrorKind::UnclosedXmlElement)),
```

Comment:
Again, I think we get this check for free if `Reader::check_end_names` is enabled via [`Error::UnexpectedEof`](https://docs.rs/quick-xml/latest/quick_xml/enum.Error.html).